app/Http/Transformers/SignupTransformer.php

```diff
@@ -49,14 +58,25 @@ public function transform(Signup $signup)
      * Include posts
      *
      * @param \Rogue\Models\Signup $signup
+     * @param \League\Fractal\ParamBag|null
      * @return \League\Fractal\Resource\Collection
      */
-    public function includePosts(Signup $signup)
+    public function includePosts(Signup $signup, ParamBag $params = null)
     {
-        // Only allow an admin or the user who owns the signup to see the signup's unapproved posts.
-        $post = $signup->visiblePosts;
+        // Validate `type` parameter, if provided:
+        if ($params) {
+            $usedParams = array_keys(iterator_to_array($params));
+
+            if ($invalidParams = array_diff($usedParams, $this->validParams)) {
+                throw ValidationException::withMessages([
+                    'include' => 'Parameters must be one of: ' . implode(',', $this->validParams),
+                ]);
+            }
 
-        return $this->collection($post, new PostTransformer);
+            return $this->collection($signup->visiblePosts($params->get('type')), new PostTransformer);
+        } else {
+            return $this->collection($signup->visiblePosts, new PostTransformer);
```

Comment:
Hmm, what error do you get when you pass `null` to that function? Also were you able to see if there was any difference in what SQL queries are made when you use the relationship property `$signup->visiblePosts` vs. the function `$signup->visiblePosts()`?

Reply:
@DFurnes when I ass `null` to the function, no posts are returned for some reason 🤔 
> Also were you able to see if there was any difference in what SQL queries are made when you use the relationship property $signup->visiblePosts vs. the function $signup->visiblePosts()?
That is interesting! Not quite sure how to see this? However, if I `dd($this->collection($signup->visiblePosts, new PostTransformer));`
I get: 
`League\Fractal\Resource\Collection {#1938
  #data: Illuminate\Database\Eloquent\Collection {#1958
    #items: array:1 [
      0 => Rogue\Models\Post {#1902
...`
`

but with `dd($this->collection($signup->visiblePosts(null), new PostTransformer));` I can't scroll all the way up but here's the beginning of what I can see:
`
                  #verbosity: 32
                  #verbosityMap: array:5 [ …5]
                  -application: Illuminate\Console\Application {#52 …18}
                  -processTitle: null
                  -aliases: []
                  -definition: Symfony\Component\Console\Input\InputDefinition {#691 …6}
                  -help: null
                  -ignoreValidationErrors: false
                  -applicationDefinitionMerged: false
                  -applicationDefinitionMergedWithArgs: false
                  -code: null
                  -synopsis: []
                  -usages: []
                  -helperSet: Symfony\Component\Console\Helper\HelperSet {#507 …2}
                  -name: "queue:failed"
                  -hidden: false
                  -description: "List all of the failed queue jobs"
                }
                "command.queue.flush" => Illuminate\Queue\Console\FlushFailedCommand {#692
                  #name: "queue:flush"
                  #description: "Flush all of the failed queue jobs"
                  #laravel: Illuminate\Foundation\Application {#565}
                  #input: null
                  #output: null
                  #signature: null
                  #hidden: false
                  #verbosity: 32
                  #verbosityMap: array:5 [ …5]
                  -application: Illuminate\Console\Application {#52 …18}
                  -processTitle: null
                  -aliases: []
                  -definition: Symfony\Component\Console\Input\InputDefinition {#693 …6}
                  -help: null
                  -ignoreValidationErrors: false
                  -applicationDefinitionMerged: false
                  -applicationDefinitionMergedWithArgs: false
                  -code: null
                  -synopsis: []
                  -usages: []
                  -helperSet: Symfony\Component\Console\Helper\HelperSet {#507 …2}
                  -name: "queue:flush"
`

Reply:
@DFurnes oof I figured it out! I was missing the `->get();` for the queries `if ($type === null)` 🤦‍♀️ !

Reply:
Be careful calling `->get()`! That tells Eloquent to run the query then and there, so by using it inside the relationship we're accidentally making a bunch of extra queries.

To see what database queries are being made, I'd recommend using Clockwork (accessible at [`/__clockwork`](http://rogue.test/__clockwork)). Just keep that open in a tab while you make requests on your local Postman, and then you can go back and inspect what happened each time:

![screen shot 2018-07-19 at 5 05 38 pm](https://user-images.githubusercontent.com/583202/42969976-fc1c4a48-8b75-11e8-8d93-be91c620c667.png)

I think we can still do a little work here - the plain `v3/signups` index makes one SQL query, but once we add the `?include=posts` query parameter it jumps to 61!! 😱 Instead of scoping the query on each individual signup here, let's use the [`scopeVisiblePosts`](https://github.com/DoSomething/rogue/blob/0446b280a57bc8a2e92f6730cbf86d09aa303b21/app/Models/Signup.php#L215-L228) scope in the controller to [eager-load](https://laravel.com/docs/5.6/eloquent-relationships#eager-loading) these before they hit the transformer.

This is a bit messy, but something to work off of - down to 4 SQL queries!

```diff
diff --git a/app/Http/Controllers/SignupsController.php b/app/Http/Controllers/SignupsController.php
index 693eaa9c..1da966fa 100644
--- a/app/Http/Controllers/SignupsController.php
+++ b/app/Http/Controllers/SignupsController.php
@@ -91,8 +91,13 @@ class SignupsController extends ApiController
         $query = $this->filter($query, $filters, Signup::$indexes);
 
         // Only allow an admin or the user who owns the signup to see the signup's unapproved posts.
-        if ($request->query('include') === 'posts') {
-            $query = $query->withVisiblePosts();
+        if (starts_with($request->query('include'), 'posts')) {
+            $types = (new \League\Fractal\Manager)
+                ->parseIncludes($request->query('include'))
+                ->getIncludeParams('posts')
+                ->get('type');
+
+            $query = $query->withVisiblePosts($types);
         }
 
         $orderBy = $request->query('orderBy');
diff --git a/app/Http/Transformers/SignupTransformer.php b/app/Http/Transformers/SignupTransformer.php
index a27c6437..72e3621d 100644
--- a/app/Http/Transformers/SignupTransformer.php
+++ b/app/Http/Transformers/SignupTransformer.php
@@ -74,9 +74,7 @@ class SignupTransformer extends TransformerAbstract
             }
         }
 
-        $type = $params ? $params->get('type') : null;
-
-        return $this->collection($signup->visiblePosts($type), new PostTransformer);
+        return $this->collection($signup->posts, new PostTransformer);
     }
 
     /**
@@ -235,13 +241,15 @@ class Signup extends Model
      *
      * @return \Illuminate\Database\Eloquent\Builder
      */
-    public function scopeWithVisiblePosts($query)
+    public function scopeWithVisiblePosts($query, $type)
     {
-        if (! is_staff_user()) {
-            return $query->with(['posts' => function ($query) {
-                 $query->where('status', 'accepted')
-                ->orWhere('northstar_id', auth()->id());
-            }]);
-        }
+        return $query->with(['posts' => function ($query) use ($type) {
+            $query->whereIn('type', $type);
+
+            if (! is_staff_user()) {
+               $query->where('status', 'accepted')
+                    ->orWhere('northstar_id', auth()->id());
+            }
+        }, 'posts.tags']);
     }
 }
```

Reply:
@DFurnes ahh that is so interesting, thank you for sharing! You mentioned clockwork to me before but I forgot about it so thank you for reminding me! Very helpful in terms of performance :) 

Also, didn't realize that about using `->get()` - I've always wondered! It is interesting though because when I did the same exact code in the previous commit before the most recent, without `->get()`, I couldn't get the results that I wanted. 

I went about the way that you suggested above (so smart!) and it seems that I am running into the same problem - the query isn't working and isn't actually sorting by type (you can see this with some of the failed tests). Do you have any ideas about this?

Reply:
Try swapping to use the plain `$signup->posts` relation in the transformer... that seems to get things properly filtering on my local. It seems like using `$signup->visiblePosts` still "forces" Laravel to re-fetch the relation.

```diff
diff --git a/app/Http/Transformers/SignupTransformer.php b/app/Http/Transformers/SignupTransformer.php
index 9569f92e..31919adb 100644
--- a/app/Http/Transformers/SignupTransformer.php
+++ b/app/Http/Transformers/SignupTransformer.php
@@ -55,10 +55,7 @@ class SignupTransformer extends TransformerAbstract
      */
     public function includePosts(Signup $signup)
     {
-        // Only allow an admin or the user who owns the signup to see the signup's unapproved posts.
-        $post = $signup->visiblePosts;
-
-        return $this->collection($post, new PostTransformer);
+        return $this->collection($signup->posts, new PostTransformer);
     }
```

Reply:
Actually wait - better yet, let's change how we eager-load in `scopeWithVisiblePosts`! That should let us keep using `$signup->visiblePosts` like we want to in `SignupTransformer`.

```diff
diff --git a/app/Models/Signup.php b/app/Models/Signup.php
index 82b20baf..662b64cc 100644
--- a/app/Models/Signup.php
+++ b/app/Models/Signup.php
@@ -219,7 +219,7 @@ class Signup extends Model
      */
     public function scopeWithVisiblePosts($query, $types = null)
     {
-        return $query->with(['posts' => function ($query) use ($types) {
+        return $query->with(['visiblePosts' => function ($query) use ($types) {

             if ($types) {
                 $query->whereIn('type', $types);
```

Reply:
@DFurnes updated to that and still not passing / working :(